source/common/event/dispatcher_impl.cc

```diff
@@ -258,23 +258,23 @@ void DispatcherImpl::updateApproximateMonotonicTimeInternal() {
 }
 
 void DispatcherImpl::runPostCallbacks() {
-  while (true) {
-    // It is important that this declaration is inside the body of the loop so that the callback is
-    // destructed while post_lock_ is not held. If callback is declared outside the loop and reused
-    // for each iteration, the previous iteration's callback is destructed when callback is
-    // re-assigned, which happens while holding the lock. This can lead to a deadlock (via
-    // recursive mutex acquisition) if destroying the callback runs a destructor, which through some
-    // callstack calls post() on this dispatcher.
-    std::function<void()> callback;
-    {
-      Thread::LockGuard lock(post_lock_);
-      if (post_callbacks_.empty()) {
-        return;
-      }
-      callback = post_callbacks_.front();
-      post_callbacks_.pop_front();
+  std::list<std::function<void()>> callbacks;
+  {
+    // Take ownership of the callbacks under the post_lock_. The lock must be released before
+    // callbacks execute. Callbacks added after this transfer will re-arm post_cb_ and will execute
+    // later in the event loop.
+    Thread::LockGuard lock(post_lock_);
+    if (post_callbacks_.empty()) {
+      return;
     }
+    callbacks = std::move(post_callbacks_);
```

Comment:
I didn't think std::move defined clearly what state `post_callbacks_` would be in after this statement. So I was thinking maybe swap would be better. WDYT?

Reply:
post_callbacks_ is a list, so I think this is well defined.  I could add post_callbacks_.clear(); after the move.

Reply:
ok that seems like it would de-risk it. From https://en.cppreference.com/w/cpp/utility/move

> Unless otherwise specified, all standard library objects that have been moved from are placed in a valid but unspecified state. That is, only the functions without preconditions, such as the assignment operator, can be safely used on the object after it was moved from: ... `str.clear(); // OK, clear() has no preconditions`

is that better than swap? I'm fine either way.

I did not see anything about std::list's move semantics in https://en.cppreference.com/w/cpp/container/list

Reply:
std::swap is defined as 3 move assignments.  Added the clear and a comment.

Reply:
Instead of the `clear()`, I think it would be better to `ASSERT(post_callbacks_.empty())`. Then if something gets messed up and it somehow became a copy instead of a move (due to c++-shenanigans or something) we'll notice quickly, instead of callbacks not getting run.

Reply:
Done.

Reply:
Just jumping in here... what about:
```c++
std::list<std::function<void()>> callbacks = [this]() {
  Thread::LockGuard lock(post_lock_);
  return std::exchange(post_callbacks_, {});
}();
```
or some variation using `std::exchange`. The pattern of using exchange is mentioned in reference to an event Dispatcher class in [an article](https://www.fluentcpp.com/2020/09/25/stdexchange-patterns-fast-safe-expressive-and-probably-underused/) on fluentcpp.com.

Reply:
I went through the article.  I think the most relevant section is the "Why not just move?".  The issue of expressing the "empty after move" constraint is addressed here by means of the ASSERT(post_callbacks_,empty());